src/utils/git.rs

```diff
@@ -267,21 +267,31 @@ pub fn fetch(
     })
 }
 
+fn get_ref_from_name<'repo>(repo: &'repo Repository, name: &str) -> Option<Reference<'repo>> {
+    match repo.find_branch(name, BranchType::Local) {
+        Ok(branch) => Some(branch.into_reference()),
+        Err(e) => { // toDo may be other errors here...
+            let tag_ref = repo.find_reference(&format!("refs/tags/{}", name)).expect("reference not found");
+            if tag_ref.is_tag() {
+                Some(tag_ref)
+            } else { // not a branch or a tag
+                None
+            }
+        }
+    }
+}
+
 /// Push the branches or tags given in ref_names
 pub fn push(repo: &Repository, remote: &mut Remote, ref_names: &[&str]) -> Result<()> {
-    let refs: Vec<String> = ref_names
-        .iter()
-        .map(|name: &&str| {
-            format!(
-                "refs/heads/{}:refs/heads/{}",
-                name.to_string(),
-                name.to_string()
-            )
-        })
-        .collect();
+    let mut full_names: Vec<String> = vec![];
+    for name in ref_names {
+        let reference = get_ref_from_name(&repo, name).expect("failed to find reference from name");
+        let full_name = String::from(reference.name().expect("failed to get full name of ref"));
```

Comment:
We shouldn't `panic` here either; basically, down in the internal modules, nothing should ever panic. It's more forgivable at the top level of a CLI application.

Reply:
638721c